### ISSUE TITLE
PayPal Plus Payment redirects to Standard Paypal instead of showing the typical Paypal Plus iFrame

### ISSUE THREAD

**Issue by ben-eSM** [isotope/core#2145]
The changes introduced in Isotope 2.6.11 and 2.6.12 seem to break the PayPal PLUS checkout.
I have multiple shops which lead the user to the standard Paypal checkout where an account is needed instead of showing the PayPal PLUS Iframe where they usually would choose their preferred payment method. Going back to 2.6.10 and it works again as intended. Seems related to the changes applied to Paypal Plus payment code.

**Comment by aschempp** [comment]
does the "standard checkout" work as well? Or is there a difference (except the redirect)?

**Comment by ben-eSM** [comment]
The standard checkout seems to work, yes. But you can't choose to pay by credit card or one of the other options of PayPal Plus without creating a login.

**Comment by aschempp** [comment]
did you turn on the option to "accept payment from non-PayPal users" in the account configuration? Also, did your account (previously) accept invoices through PayPal PLUS?

**Comment by ben-eSM** [comment]
It also affected clients who already received lots of PayPal Plus payments before. After the update to 2.6.12, the payment method only serves the standard PayPal checkout. The accounts seem to be setup properly. But I guess otherwise it shouldn't have worked before if the accounts would be configured incorrectly.

**Comment by aschempp** [comment]
I don't know about that. The issue fixed in 2.6.12 was for where clients had everything set up correctly (but probably not signed up for PayPal PLUS), and they just got a blank page (no iframe content) from PayPal. Do you have any client **with** PayPal PLUS where the iframe still works correctly? Did your clients manually sign up for PayPal PLUS? I'm asking because it _looks like_ PayPal changed their API response/behavior, that's why I changed it in 2.6.12

**Comment by ben-eSM** [comment]
Ich have two clients with PayPal Plus. Both Shops show the standard checkout instead of the iframe after updating to 2.6.12. Both clients did manually sign up for PayPal Plus.

**Comment by aschempp** [comment]
Can you please enable debug logging and paste the paypal response when initiating a transaction (you might need to anonymize the data)

**Comment by ben-eSM** [comment]
This is the response content of the paypal log file:

```
[29-May-2020 14:43:43] {"id":"PAYID-L3IQG7Y7NG44706B6433705R","intent":"sale","state":"created","payer":{"payment_method":"paypal","payer_info":{"email":"xx@xx.xx","first_name":"xxx","last_name":"xxx","billing_address":{"line1":"xxx. 99","city":"xxxx","postal_code":"12345","country_code":"DE"}}},"potential_payer_info":{"billing_address":{"line1":"xxx. 99","line2":"","city":"xxxx","state":"","postal_code":"12345","country_code":"DE"}},"transactions":[{"amount":{"total":"99.99","currency":"EUR"},"item_list":{"items":[{"name":"yyy","sku":"yyy.yy.yyy","price":"99.99","currency":"EUR","quantity":1},{"name":"Versand mit DHL","price":"999","currency":"EUR","quantity":1}]},"related_resources":[]}],"create_time":"2020-05-29T12:43:43Z","links":[{"href":"https://api.paypal.com/v1/payments/payment/PAYID-L3IQG7Y7NG44706B6433705R","rel":"self","method":"GET"},{"href":"https://www.paypal.com/cgi-bin/webscr?cmd=_express-checkout&token=EC-7EH96642NA357413W","rel":"approval_url","method":"REDIRECT"},{"href":"https://api.paypal.com/v1/payments/payment/PAYID-L3IQG7Y7NG44706B6433705R/execute","rel":"execute","method":"POST"}]}
```

**Comment by aschempp** [comment]
There relevant part are the links at the bottom:

```json
  "links": [
    {
      "href": "https://api.paypal.com/v1/payments/payment/PAYID-L3IQG7Y7NG44706B6433705R",
      "rel": "self",
      "method": "GET"
    },
    {
      "href": "https://www.paypal.com/cgi-bin/webscr?cmd=_express-checkout&token=EC-7EH96642NA357413W",
      "rel": "approval_url",
      "method": "REDIRECT"
    },
    {
      "href": "https://api.paypal.com/v1/payments/payment/PAYID-L3IQG7Y7NG44706B6433705R/execute",
      "rel": "execute",
      "method": "POST"
    }
  ]
```

Originally, the PayPal docs said we need to use the `approval_url` for the iframe. As that did not work for my client, I looked at the `method`, and assumed `REDIRECT` means - just that.

Can you try to adjust the payment method to use `self` URL instead and see if the iframe works with that? Or can you find any hint to that in the PayPal docs?

**Comment by andreas361** [comment]
@ben-eSM Have you alreay tested what @aschempp suggested? We have the same problem in a shop of our client.

**Comment by operatorone** [comment]
I tried it in my instance too and I'm also redirected. Can I do something to help to find the error? @aschempp You still have access to my project to test this if you wish.

**Comment by andreas361** [comment]
Any news on this?

**Comment by timobanaszak** [comment]
Just want to dive in: Had the same problem and the downgrade to 2.6.10 fixed the issue. Any way I can contribute to find a solution?

**Comment by asaage** [comment]
Same problem here 😐 - the iframe is gone and the redirect offers no plus options.

**Comment by andreas361** [comment]
Anybody news on this (besides downgrade)?

**Comment by aschempp** [comment]
> Can you try to adjust the payment method to use `self` URL instead and see if the iframe works with that? Or can you find any hint to that in the PayPal docs?

There won't be any news until someone can debug this and provide any useful feedback. As I do not run any shop - nor do I use PayPal PLUS - theres not much I can do except wait for feedback or someone using our commercial support option (https://circle.isotopeecommerce.org) to let me debug a real shop.

**Comment by 0xhtml** [comment]
On [this](https://developer.paypal.com/docs/paypal-plus/germany/integrate/create-payment-resource/#sample-response) page of the PayPal PLUS docs an example response is shown in which the method for the approval_url is 'REDIRECT'. However, the docs don't state that the user should be redirected to the approval_url, but rather state [here](https://developer.paypal.com/docs/paypal-plus/germany/integrate/integrate-payment-wall/#embed-the-iframe) that the iframe should be generated with the approval_url.

Downgrading to version 2.6.10 solves this issue for multiple people including me. Hence, I conclude that removing lines 54-56 in 
core/system/modules/isotope/library/Isotope/Model/Payment/PaypalPlus.php should fix the issue.
https://github.com/isotope/core/blob/15b1376c5ba76c27e10306bf7ebc767ca20aa735/system/modules/isotope/library/Isotope/Model/Payment/PaypalPlus.php#L54-L56 
These lines were added with commit b1bd72c5d3b67abc198cf45cde4da313208fec48 (released in version 2.6.11) to redirect users to the standard PayPal checkout if PayPal PLUS isn't available. For reference: https://github.com/isotope/core/issues/2145#issuecomment-637358874

**Comment by aschempp** [comment]
Fixed in 73d8470cb2e030279033bd9835407730a4a23396
